src/pybind/mgr/prometheus/module.py

```diff
@@ -187,15 +188,24 @@ class MetricCollectionThread(threading.Thread):
     def __init__(self, module):
         # type: (Module) -> None
         self.mod = module
+        self.active = True
         super(MetricCollectionThread, self).__init__(target=self.collect)
 
     def collect(self):
         self.mod.log.info('starting metric collection thread')
-        while True:
+        while self.active:
             self.mod.log.debug('collecting cache in thread')
             if self.mod.have_mon_connection():
                 start_time = time.time()
-                data = self.mod.collect()
+
+                try:
+                    data = self.mod.collect()
+                except Exception as e:
+                    # Log any issues encountered during the data collection and continue
+                    self.mod.log.warn(str(traceback.format_exc()))
```

Comment:
```suggestion
                    self.mod.log.excepion("failed to collect")
```

This already prints the full traceback into the log.

Reply:
I have amended the commit to use `self.mod.log.exception` instead.